streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -339,28 +376,38 @@ private void closeTopology() {
 
     /**
      * <pre>
-     *  - {@link #suspend()}
-     *    - close topology
-     *    - {@link #commit()}
-     *      - flush state and producer
-     *      - write checkpoint
-     *      - commit offsets
-     *  - close state
+     * - {@link #suspend(boolean) suspend(clean)}
+     *   - close topology
+     *   - if (clean) {@link #commit(boolean) commit(noNewTransaction)}
+     *     - flush state and producer
+     *     - commit offsets
+     * - close state
+     *   - if (clean) write checkpoint
+     * - if (eos) close producer
      * </pre>
+     * @param clean shut down cleanly (ie, incl. flush and commit) if {@code true} --
+     *              otherwise, just close open resources
      */
     @Override
-    public void close() {
+    public void close(final boolean clean) {
         log.debug("{} Closing", logPrefix);
         try {
-            suspend();
-            closeStateManager(true);
+            suspend(clean);
+            closeStateManager(clean);
```

Comment:
This is not introduce in this PR, but I'm still not sure if it is safe to call `closeStateManager` multiple times, since we are trying to unlock the directory at the end, and hence if unlocking itself throw an exception, we should not try to call `closeStateManager` again. I'd suggest making the try-catch block finer grained in this case, e.g. if we expect only capturing exceptions from `suspend` then we should not cover `partitionGroup.close()` and `metrics.removeAllSensors`, and also `closeStateManager` would better be called only once depending on whether suspend succeeded or failed.